### ISSUE TITLE
System.Net.WebSockets.Client - ReceiveAsync multiple-call failure behavior platform dependent.

### ISSUE THREAD

**Issue by Clockwork-Muse** [dotnet/corefx#13169]
As I chronicle at one point in #12487 , `ReceiveAsync` has very different failure behaviors between Windows and all other platforms (potentially to other Windows platforms, too).  Note this behavior is **not** the result of a race condition.
I expect that `SendAsync` is also affected by this, but haven't investigated fully.

The critical part for Windows is in [WinHttpWebSocket](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs#L395-L446), but what ends up happening is that when the second call to `ReceiveAsync` happens, the second call sets up the exception on the TCS returned by the _first_ call, then also returns that **same** task:
```c#
if (receiveOperationAlreadyPending)
{
    var exception = new InvalidOperationException(
    SR.Format(SR.net_Websockets_AlreadyOneOutstandingOperation, "ReceiveAsync"));
    _operation.TcsReceive.TrySetException(exception);

    Abort();
}

return _operation.TcsReceive.Task;
```
(Given the locking that occurs, this is guaranteed to happen even if submitted on two concurrently running threads, although which one gets it would be up in the air)

This is clearly different from the behavior on Linux/elsewhere, when going through [ManagedWebSocket](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ManagedWebSocket.cs#L288-L309) it returns a separate task containing an `InvalidOperationException`, leaving the first task to complete "on it's own" - it dies as `Abort`ed, I believe throwing a `WebSocketException`:
```c#
try
{
    WebSocketValidate.ThrowIfInvalidState(_state, _disposed, s_validReceiveStates);

    Debug.Assert(!Monitor.IsEntered(StateUpdateLock), $"{nameof(StateUpdateLock)} must never be held when acquiring {nameof(ReceiveAsyncLock)}");
    lock (ReceiveAsyncLock) // synchronize with receives in CloseAsync
    {
        ThrowIfOperationInProgress(_lastReceiveAsync);
        Task<WebSocketReceiveResult> t = ReceiveAsyncPrivate(buffer, cancellationToken);
        _lastReceiveAsync = t;
        return t;
    }
}
catch (Exception exc)
{
    return Task.FromException<WebSocketReceiveResult>(exc);
}
```

.... sooooo, what's the expected behavior here?  What should the tests anticipate?

**Comment by Clockwork-Muse** [comment]
cc @stephentoub , @CIPop

**Comment by karelz** [comment]
Affects both OSs Linux and Windows

**Comment by davidsh** [comment]
@Clockwork-Muse Can you please summarize the bug here?  I read the issue here but I don't understand what behavior you expect vs. what you observe?

**Comment by Clockwork-Muse** [comment]
I believe the primary problem is(/was?) that on Linux you have two different tasks for two different invocations, which yield different results.

On Windows, because there's only one task used for both calls, and the Task itself is set to an exception, both parties that wait on it see the same end state.  
On Linux, that's not the case.  The second call to `ReceiveAsync` dies with the expected error - that there's already an operation waiting, however the first call, the one that's doing the actual waiting, does not.  I'd have to re-dive, but I'm not completely sure it's guaranteed to die at all, and instead may actually return results.  However, I believe it often dies as if the connection was aborted, or canceled independently.  Which would make debugging a right pain.... (phantom task cancellations).

Now, I don't care so much about returning the same task, but what I _expect_ is that on Linux, like on Windows, both calls to `ReceiveAsync` yield the same exception, with the same message; that they died because somebody called it a second time.  
That, or change the class so the first request is guaranteed to succeed, and the second one doesn't disrupt that at all; the documentation only states only one call at a time is supported.  Although having both die does make debugging easier.

**Comment by davidsh** [comment]
>Now, I don't care so much about returning the same task, but what I expect is that on Linux, like on Windows, both calls to ReceiveAsync yield the same exception, with the same message; that they died because somebody called it a second time.

Sounds like we have  Windows vs. Linux behavior difference that needs investigating.

cc: @stephentoub

**Comment by karelz** [comment]
cc @Priya91

**Comment by karelz** [comment]
@stephentoub can you please help?

**Comment by stephentoub** [comment]
The Linux behavior here matches the desktop behavior; it's the .NET Core on Windows implementation that's the odd-man-out.

Consider this short repro:
```C#
using System;
using System.Net.WebSockets;
using System.Threading;
using System.Threading.Tasks;

class Program
{
    static void Main() => MainAsync().GetAwaiter().GetResult();
    static async Task MainAsync()
    {
        var cws = new ClientWebSocket();
        await cws.ConnectAsync(new Uri("ws://corefx-net.cloudapp.net/WebSocket/EchoWebSocket.ashx"), CancellationToken.None);
        Task t1 = cws.ReceiveAsync(new ArraySegment<byte>(new byte[1]), CancellationToken.None);
        Task t2 = cws.ReceiveAsync(new ArraySegment<byte>(new byte[1]), CancellationToken.None);
        Thread.Sleep(1000);
        Console.WriteLine(t1.Exception.InnerException.GetType());
        Console.WriteLine(t2.Exception.InnerException.GetType());
    }
}
```

On desktop:
```
System.Net.WebSockets.WebSocketException
System.InvalidOperationException
```

On .NET Core on Linux:
```
System.Net.WebSockets.WebSocketException
System.InvalidOperationException
```

On .NET Core on Windows:
```
System.InvalidOperationException
System.InvalidOperationException
```

**Comment by stephentoub** [comment]
(Though it appears https://github.com/dotnet/corefx/pull/17444 changed this behavior... my tests above were on .NET Core 1.1.... checking on .NET Core 2.0...)

**Comment by stephentoub** [comment]
Now as of #17444, on .NET Core on both Windows and Unix, `t1` ends as Canceled and `t2` ends with an InvalidOperationException.  So they're consistent with each other, but not with desktop.  But, @davidsh, from your comments in https://github.com/dotnet/corefx/issues/5200#issuecomment-288525147, it sounds like this is the desired behavior and this issue can just be closed?

**Comment by davidsh** [comment]
>it sounds like this is the desired behavior and this issue can just be closed?

Does "it" means that "An OperationCanceledException" is thrown if the async I/O is canceled via an explicit request, i.e. cancellationtoken being set or via timer".

If so, then yes, that is the desired behavior which matches the .NET guidelines for async best-practices

**Comment by stephentoub** [comment]
> Does "it" mean that "An OperationCanceledException" is thrown if the async I/O is canceled via an explicit request, i.e. cancellationtoken being set or via timer".

That's not exactly the situation here.  The situation here isn't an explicit request to cancel or timeout: it's using the API incorrectly, issuing a second ReceiveAsync when there's already one pending, in which case the first one is completing as canceled.

**Comment by davidsh** [comment]
I think that is still ok given the idea that an action was taken to cause a pending operation to cancel.  Folks like @terrajobst are other folks that could comment on the "correctness" of the exception being thrown.

**Comment by stephentoub** [comment]
Ok.  I'll consider the updated behavior by-design then and this issue is now addressed as the behavior is consistent between Windows and Unix.
